### PR TITLE
Add native TR-HASH model support

### LINKED ISSUE #48193: Add native support for TR-HASH deterministic token-routed MoE
### Model description

We are the model authors and would like native Transformers support for TR-HASH MoE so the public checkpoint can load with `AutoModelForCausalLM.from_pretrained(...)` without `trust_remote_code=True`.

TR-HASH is a 201.2M-parameter decoder-only causal language model with 16 layers, GQA (14 query heads / 2 KV heads), tied embeddings, four stored experts with top-2 activation, and an always-on shared SwiGLU path.

Unlike conventional sparse MoE models, expert selection is not produced by a learned router. It is a deterministic function of token IDs, the layer index, and persisted per-layer multi-hash routing tables.

The principal implementation differences are:

1. The MLP receives token IDs in addition to hidden states.
2. Routing tables are persisted integer buffers and must stay paired with expert weights.
3. Expert projections are stored as stacked tensors:
   - `expert_gate`: `[num_experts, hidden_size, expert_width]`
   - `expert_up`: `[num_experts, hidden_size, expert_width]`
   - `expert_down`: `[num_experts, expert_width, hidden_size]`
4. A shared SwiGLU branch is always active alongside the routed experts.
5. The public checkpoint preserves native tensor names without a weight conversion.

A self-contained `PreTrainedModel` / `PretrainedConfig` adapter already exists and has been validated on the released 201.2M SFT checkpoint. It supports modern and legacy KV caches, generation, and `AutoConfig` / `AutoModelForCausalLM` through `auto_map`.

Validation completed:

- exact full-logit parity with the reference PyTorch implementation
- exact cached-step logit parity with the reference cache path
- greedy generation verified
- Transformers 4.30 and 5.14 tested
- 73 focused framework/model tests passed, 3 skipped
- Ruff check and format check passed

Native support would unblock standard evaluation, Intel AutoRound quantization, and downstream serving integrations that reject unknown `model_type` values or remote code.

Expected native loading:

```python
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained(
    "AETHORIA-AI/TR-HASH-MoE-200M-160B-SFT"
)
```

Definition of done:

1. `AutoConfig` recognizes `tr_hash_moe`.
2. `AutoModelForCausalLM` loads the public checkpoint without remote code.
3. Persisted routing tables and stacked 3D expert weights load without conversion.
4. Eager and cached logits match the reference implementation within the agreed tolerance.
5. Greedy generation matches the reference token sequence.
6. The architecture is documented and covered by standard model tests.

We can provide reference logits, routing traces, small fixtures, and any additional architectural details maintainers need.

### Open source status

- [x] The model implementation is available
- [x] The model weights are available

### Provide useful links for the implementation

- Public SFT checkpoint (Apache-2.0): https://huggingface.co/AETHORIA-AI/TR-HASH-MoE-200M-160B-SFT
- Reference implementation: https://github.com/Complexity-ML/complexity-framework
- Validated Transformers adapter: https://github.com/Complexity-ML/complexity-framework/pull/9
- Current serving engine: https://github.com/Complexity-ML/TR-Hash-i64
- Related vLLM model-support request: https://github.com/vllm-project/vllm/issues/53286
- Authors: @Complexity-ML / AETHORIA-AI

### PR DESCRIPTION
**Pull request by Complexity-ML**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48194) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48194)

## What does this PR do?

Adds native Transformers support for the TR-HASH deterministic token-routed MoE architecture and the public `AETHORIA-AI/TR-HASH-MoE-200M-160B-SFT` checkpoint.

TR-HASH differs from learned-router MoEs in three important ways:

- expert selection is a deterministic function of token IDs, layer index, and persisted multi-hash route tables;
- each token activates two of four stored experts plus an always-on shared SwiGLU branch;
- routed gate/up/down weights are stored as stacked 3D parameters.

This PR adds:

- `TRHashConfig`, `TRHashModel`, and `TRHashForCausalLM`;
- `AutoConfig`, `AutoModel`, and `AutoModelForCausalLM` mappings;
- eager and SDPA attention through the standard Transformers attention interface;
- dynamic-cache generation, gradient checkpointing, tied embeddings, persisted routing metadata, and standard save/load support;
- model documentation and a slow public-checkpoint integration test.

`inputs_embeds`-only generation is intentionally unsupported because deterministic expert routing requires the original token IDs.

## Validation

- Model suite: **134 passed, 169 skipped, 727 subtests passed**.
- Repository checks: `check_repo`, `check_auto`, `check_inits`, `check_doc_toc`, `check_config_attributes`, `check_config_docstrings`, `check_modeling_structure`, `check_modeling_rules_doc`, Ruff, and `git diff --check` pass.
- The public 290-tensor checkpoint loads strictly with no missing or unexpected keys.
- Fixed-prompt SDPA logits match the reference Complexity Framework implementation exactly (`max_abs_error = 0.0`).
- Multi-token cache continuation stays within `2.58e-05` maximum absolute error and produces identical argmax tokens.

We are the model authors and can provide additional routing traces or reference logits if useful during review.

Closes #48193

<details>
<summary>🤖 mlinter review state</summary>

- hash: `1f798562`
- commit: `1a279448`
- review: https://github.com/huggingface/transformers/pull/48194#pullrequestreview-5065535513

</details>

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
hey! before we can review the pr, can you follow the contribution standards and standard the architecture comparing with th others we supports? we have a whole lot of guides on how to create a modular file.

**Comment by Complexity-ML** [comment]
Thanks for the guidance — I refactored the implementation to follow the modular Transformers architecture.

- Added `modular_tr_hash_moe.py` and regenerated the configuration/modeling files with the official modular converter.
- Reused Qwen3 only as the standard modular GQA/decoder skeleton; the generated TR-HASH runtime is standalone.
- Kept TR-HASH's native deterministic MoE engine: persisted multi-hash routing tables, stacked 3D expert tensors, top-2 routing, and the always-on shared SwiGLU path.
- Normalized the module and `model_type` to `tr_hash_moe`, matching the released public checkpoint.

Validation completed locally:
- official modular conversion and repository structure checks pass
- auto mappings, init/docs/copy/date checks pass
- Ruff check and format checks pass
- 136 model tests and 727 subtests pass (167 skipped)
- the public checkpoint integration test passes
- native implementation and the current Hub adapter produce bit-exact logits on the public checkpoint (`max_abs_diff = 0.0`), including the cached path

The refactor is pushed in `07dda1e131`. Please let me know if you would like any further decomposition or targeted tests before detailed review.

**Review by ArthurZucker — COMMENTED** [review]
Hey! Your Hash engine is great for routing, I love it! The rest can almost entirely be inherited. The hash is a router. The rest, a shared expert, the gate up down can be fused, the pair metadata can be serialized one in the weights + conversion script as they are fixed, the attention needs to be fully inherited, removing the flag if you have a single model! 

Let's adapt the naming, engine -> router! 

The rotary embedding needs to be the same as llama, as it just is! 

The rms norm is very similar to all we have, but alright if new. 
There are too many validations, and too many "types" the defaults are not type remote the arg if it supports a unique value.

**Comment by Complexity-ML** [comment]
Thanks for the direction — I refactored the model around the standard Transformers components in `6f8adec45c`.

- renamed the hash component from `engine` to `router` and adopted the standard router output contract
- inherited Qwen3 attention, Llama-compatible RoPE, RMSNorm, decoder/cache behavior, and the shared SwiGLU MLP
- moved routed experts to the standard fused `gate_up_proj` / `down_proj` layout and added checkpoint conversion from the released separate 3D tensors
- kept the persisted route table as the canonical routing metadata; the legacy redundant pair/code buffers are ignored during native loading
- removed single-value architecture flags and bespoke validation branches from the config
- added standard TP/EP plans for the fused MoE experts

Validation on the released public checkpoint:

- 0 missing, unexpected, or mismatched keys
- exact reference token prediction preserved: `[13825, 12, 265, 202, 17, 17, 7, 224]`
- 151 focused tests passed, 149 skipped, and 651 subtests passed
- the slow public-checkpoint integration test passes offline
- modular conversion, targeted Ruff checks, formatting, and `git diff --check` pass

The modular implementation is now smaller and delegates the non-routing behavior to existing Transformers components.

File: src/transformers/models/tr_hash_moe/modular_tr_hash_moe.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+        "attentions": TRHashAttention,
+    }
+    _keys_to_ignore_on_load_unexpected = [r"mlp\.engine\.fused_(expert_pairs|route_codes)"]
+
+    # trf-ignore: TRF018 (stacked expert parameters are not Linear modules)
+    def _init_weights(self, module):
+        PreTrainedModel._init_weights(self, module)
+        if isinstance(module, TRHashRouter):
+            route_table = _build_multi_hash_route_table(module.config, module.layer_idx)
+            module.route_table.copy_(route_table.to(module.route_table.device))
```
Comment (Complexity-ML) [inline src/transformers/models/tr_hash_moe/modular_tr_hash_moe.py]: Fixed in `4a681ae7c9`: replaced the in-place buffer copy with `init.copy_`. The model linter now passes.

File: src/transformers/models/tr_hash_moe/modular_tr_hash_moe.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+            remaining_weight = (1.0 - module.config.top_k_primary_weight) / (module.top_k - 1)
+            route_weights = (module.config.top_k_primary_weight, *((remaining_weight,) * (module.top_k - 1)))
+            module.route_weights.copy_(torch.tensor(route_weights, device=module.route_weights.device))
```
Comment (Complexity-ML) [inline src/transformers/models/tr_hash_moe/modular_tr_hash_moe.py]: Fixed in `4a681ae7c9`: replaced the in-place buffer copy with `init.copy_`. The model linter now passes.

File: src/transformers/models/tr_hash_moe/modular_tr_hash_moe.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+                if isinstance(module, TRHashRouter):
+                    module.resize_route_table(self.config.vocab_size)
+        return embeddings
+
+
+@auto_docstring
+class TRHashModel(TRHashPreTrainedModel, Qwen3Model):
+    def __init__(self, config: TRHashConfig):
+        super().__init__(config)
+        self.layers = nn.ModuleList(
```
Comment (Complexity-ML) [inline src/transformers/models/tr_hash_moe/modular_tr_hash_moe.py:326]: Fixed in `4a681ae7c9`: `TRHashDecoderLayer` now explicitly inherits `GradientCheckpointingLayer`. The model linter now passes.
